File: medium/018_4Sum/4sum.py
class Solution(object):
    def do_2sum(self, result, a, target, saved):
        i = 0
        j = len(a)-1
        while i < j:
            if i > 0 and a[i] == a[i-1]:
                i += 1
                continue
            if j < (len(a)-1) and a[j] == a[j+1]:
                j -= 1
                continue
            total = a[i] + a[j]
            if total == target:
                result.append(saved + [a[i], a[j]])
                i += 1
                j -= 1
            elif total < target:
                i += 1
            else:
                j -= 1

    def do_3sum(self, result, a, target, saved):
        for i in xrange(0, len(a)-2):
            if i > 0 and a[i] == a[i-1]:
                continue
            new_saved = saved + [a[i]]
            self.do_2sum(result, a[i+1:], target-a[i], new_saved)

    def fourSum(self, a, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        a.sort()
        result = []
        for i in xrange(0, len(a)-3):
            if i > 0 and a[i] == a[i-1]:
                continue
            self.do_3sum(result, a[i+1:], target-a[i], [a[i]])
        return result

def test_4sum(solution, a, target):
    result = solution.fourSum(a, target)
    print "4sum for {}".format(a)
    print result

solution = Solution()
test_4sum(solution, [1, 0, -1, 0, -2, 2], 0)
test_4sum(solution, [0, 0, 0, 0, 0, 0, 0], 0)

File: medium/018_4Sum/4sum.go
package main

import "fmt"
import "sort"

func twoSum(nums []int, target int) [][]int {
	results := make([][]int, 0)
	i, j := 0, len(nums)-1
	for i < j {
		if i > 0 && nums[i] == nums[i-1] {
			i++
			continue
		}
		if j < (len(nums)-1) && nums[j] == nums[j+1] {
			j--
			continue
		}
		sum := nums[i] + nums[j]
		if sum < target {
			i++
		} else if sum > target {
			j--
		} else {
			results = append(results, []int{nums[i], nums[j]})
			i++
			j--
		}
	}
	return results
}

func threeSum(nums []int, target int) [][]int {
	results := make([][]int, 0)
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		res := twoSum(nums[i+1:], target-nums[i])
		if len(res) == 0 {
			continue
		}
		for _, a := range res {
			b := []int{nums[i]}
			b = append(b, a...)
			results = append(results, b)
		}
	}
	return results
}

func fourSum(nums []int, target int) [][]int {
	results := make([][]int, 0)
	sort.Ints(nums)
	for i := 0; i < len(nums)-3; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		res := threeSum(nums[i+1:], target-nums[i])
		if len(res) == 0 {
			continue
		}
		for _, a := range res {
			b := []int{nums[i]}
			b = append(b, a...)
			results = append(results, b)
		}
	}
	return results
}

func testFourSum(nums []int, target int) {
	res := fourSum(nums, target)
	fmt.Printf("target %d, %v, get %v\n", target, nums, res)
}

func main() {
	testFourSum([]int{}, 0)
	testFourSum([]int{1}, 0)
	testFourSum([]int{0, 0, 0, 0, 0, 0}, 0)
	testFourSum([]int{1, 0, -1, 0, -2, 2}, 0)
}
